File: agent/accessibility.mjs
import { readFileSync, readdirSync } from "node:fs";
import { fileURLToPath } from "node:url";
import path from "node:path";
import { loadAppManifest } from "./apps.mjs";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(HERE, "..");
const SKIP_DIRECTORIES = new Set([".build", ".git", ".swiftpm", "dist", "node_modules", "test-results"]);

function filesBelow(root, extension) {
  const files = [];
  const pending = [root];
  while (pending.length) {
    const directory = pending.pop();
    for (const entry of readdirSync(directory, { withFileTypes: true })) {
      if (entry.isDirectory()) {
        if (!SKIP_DIRECTORIES.has(entry.name)) pending.push(path.join(directory, entry.name));
      } else if (entry.isFile() && entry.name.endsWith(extension)) {
        files.push(path.join(directory, entry.name));
      }
    }
  }
  return files.sort();
}

function lineNumber(content, offset) {
  let line = 1;
  for (let index = 0; index < offset; index += 1) {
    if (content.charCodeAt(index) === 10) line += 1;
  }
  return line;
}

function matches(content, expression, file) {
  return [...content.matchAll(expression)].map((match) => ({
    value: match[1],
    file,
    line: lineNumber(content, match.index || 0),
  }));
}

function sourceInventory(manifest) {
  const modifiers = [];
  const dynamicPrefixes = new Set();
  let scannedFiles = 0;
  for (const repository of manifest.repositories) {
    for (const file of filesBelow(repository.root, ".swift")) {
      scannedFiles += 1;
      const content = readFileSync(file, "utf8");
      const explicit = matches(content, /\.accessibilityIdentifier\s*\(\s*["']([^"']+)["']\s*\)/g, file)
        .map((item) => ({ ...item, repository: repository.root }));
      modifiers.push(...explicit);
      if (content.includes(".accessibilityIdentifier")) {
        for (const match of content.matchAll(/["']([a-z][a-z0-9-]*(?:\.[a-zA-Z0-9_-]+)*\.)\\\(/g)) {
          dynamicPrefixes.add(match[1]);
        }
      }
    }
  }
  return { modifiers, dynamicPrefixes, scannedFiles };
}

function specFiles(manifest) {
  const basenames = new Set(Object.values(manifest.surfaces).map((surface) => path.basename(surface.spec)));
  return filesBelow(path.join(ROOT, "packages"), ".ts").filter((file) => basenames.has(path.basename(file)));
}

function specInventory(manifest) {
  const references = [];
  const forbidden = [];
  const prefix = `${manifest.appId}.`;
  const files = specFiles(manifest);
  for (const file of files) {
    const content = readFileSync(file, "utf8");
    for (const match of matches(content, /["']([a-z][a-z0-9-]*(?:\.[a-zA-Z0-9_-]+)+)["']/g, file)) {
      if (match.value.startsWith(prefix)) references.push(match);
    }
    for (const match of content.matchAll(/\$\(\s*(["'])([^"']+)\1\s*\)/g)) {
      const selector = match[2];
      const stable = selector.startsWith("~")
        || (/identifier/i.test(selector) && !/(?:label|name|text)\s*(?:=|contains)/i.test(selector))
        || selector.startsWith("[data-testid=");
      if (!stable) {
        forbidden.push({
          kind: "text-selector",
          selector,
          file,
          line: lineNumber(content, match.index || 0),
        });
      }
    }
  }
  return { references, forbidden, files };
}

function duplicateIdentifiers(modifiers) {
  const locations = new Map();
  for (const item of modifiers) {
    const key = `${item.repository}\0${item.value}`;
    const entry = locations.get(key) || { identifier: item.value, locations: [] };
    entry.locations.push({ file: item.file, line: item.line });
    locations.set(key, entry);
  }
  return [...locations.values()]
    .filter((entry) => new Set(entry.locations.map((location) => location.file)).size > 1)
    .map((entry) => ({ kind: "duplicate-identifier", ...entry }));
}

export function validateAccessibility(appId) {
  const manifest = loadAppManifest(appId);
  const source = sourceInventory(manifest);
  const specs = specInventory(manifest);
  const defined = new Set(source.modifiers.map((item) => item.value).filter((value) => !value.includes("\\(")));
  const missing = specs.references
    .filter((item, index, all) => (
      !defined.has(item.value)
      && ![...source.dynamicPrefixes].some((prefix) => item.value.startsWith(prefix))
      && all.findIndex((other) => other.value === item.value) === index
    ))
    .map((item) => ({
      kind: "missing-identifier",
      identifier: item.value,
      referencedAt: { file: item.file, line: item.line },
    }));
  const errors = [
    ...duplicateIdentifiers(source.modifiers),
    ...missing,
    ...specs.forbidden,
  ];
  const referenced = new Set(specs.references.map((item) => item.value));
  const unused = [...new Set(source.modifiers.map((item) => item.value))]
    .filter((identifier) => identifier.startsWith(`${appId}.`) && !referenced.has(identifier))
    .sort();
  return {
    schemaVersion: 1,
    appId,
    ok: errors.length === 0,
    sourceFiles: source.scannedFiles,
    specFiles: specs.files,
    identifiers: {
      defined: defined.size,
      explicit: source.modifiers.length,
      referenced: referenced.size,
      unused,
    },
    errors,
  };
}

File: probierz-rs/src/authoring/source/accessibility.rs
use serde_json::json;
use crate::authoring::*;

/// Static source/spec accessibility audit. Its JSON shape is the former JS contract.
pub fn validate_accessibility(harness: &Path, app_id: &str) -> Result<JsonValue, Failure> {
    let loaded = manifest::load(harness, app_id)?;
    let repositories = loaded
        .document
        .get("repositories")
        .and_then(YamlValue::as_sequence)
        .cloned()
        .unwrap_or_default();
    let mut modifiers = Vec::new();
    let mut prefixes = BTreeSet::new();
    let mut scanned_files = 0usize;
    for repository in &repositories {
        let root = PathBuf::from(yaml_string(repository, "root").unwrap_or_default());
        for file in files_below(&root, ".swift")? {
            scanned_files += 1;
            let content = fs::read_to_string(&file)?;
            modifiers.extend(explicit_accessibility_identifiers(&content, &file, &root));
            if content.contains(".accessibilityIdentifier") {
                prefixes.extend(swift_dynamic_prefixes(&content));
            }
        }
    }

    let basenames: BTreeSet<String> = loaded
        .document
        .get("surfaces")
        .and_then(YamlValue::as_mapping)
        .into_iter()
        .flat_map(|surfaces| surfaces.values())
        .filter_map(|surface| yaml_string(surface, "spec"))
        .filter_map(|spec| {
            Path::new(spec)
                .file_name()
                .and_then(OsStr::to_str)
                .map(str::to_string)
        })
        .collect();
    let mut spec_files = files_below(&harness.join("packages"), ".ts")?;
    spec_files.retain(|file| {
        file.file_name()
            .and_then(OsStr::to_str)
            .is_some_and(|name| basenames.contains(name))
    });

    let mut references = Vec::new();
    let mut forbidden = Vec::new();
    let app_prefix = format!("{app_id}.");
    for file in &spec_files {
        let content = fs::read_to_string(file)?;
        for (value, offset) in quoted_values(&content) {
            if value.starts_with(&app_prefix) && valid_identifier(&value, true) {
                references.push(json!({ "value": value, "file": file.to_string_lossy(), "line": line_number(&content, offset) }));
            }
        }
        let mut cursor = 0;
        while let Some(relative) = content[cursor..].find("$(") {
            let start = cursor + relative;
            let mut index = start + 2;
            while content
                .as_bytes()
                .get(index)
                .is_some_and(u8::is_ascii_whitespace)
            {
                index += 1;
            }
            if let Some((selector, end)) = literal_at(&content, index) {
                let mut close = end;
                while content
                    .as_bytes()
                    .get(close)
                    .is_some_and(u8::is_ascii_whitespace)
                {
                    close += 1;
                }
                if content.as_bytes().get(close) == Some(&b')') && !selector.contains(['\'', '"']) {
                    let stable = selector.starts_with('~')
                        || selector.starts_with("[data-testid=")
                        || (selector.to_ascii_lowercase().contains("identifier") && {
                            let lower = selector.to_ascii_lowercase();
                            !(lower.contains("label=")
                                || lower.contains("name=")
                                || lower.contains("text=")
                                || lower.contains("label contains")
                                || lower.contains("name contains")
                                || lower.contains("text contains"))
                        });
                    if !stable {
                        forbidden.push(json!({ "kind": "text-selector", "selector": selector, "file": file.to_string_lossy(), "line": line_number(&content, start) }));
                    }
                }
                cursor = end;
            } else {
                cursor = index.saturating_add(1);
            }
        }
    }

    let defined: BTreeSet<String> = modifiers
        .iter()
        .filter_map(|item| item.get("value").and_then(JsonValue::as_str))
        .filter(|value| !value.contains("\\("))
        .map(str::to_string)
        .collect();
    let mut missing = Vec::new();
    let mut seen_missing = BTreeSet::new();
    for item in &references {
        let value = item["value"].as_str().unwrap_or_default();
        if !defined.contains(value)
            && !prefixes.iter().any(|prefix| value.starts_with(prefix))
            && seen_missing.insert(value.to_string())
        {
            missing.push(json!({ "kind": "missing-identifier", "identifier": value, "referencedAt": { "file": item["file"], "line": item["line"] } }));
        }
    }
    let mut locations: BTreeMap<(String, String), Vec<JsonValue>> = BTreeMap::new();
    for item in &modifiers {
        locations
            .entry((
                item["repository"].as_str().unwrap_or_default().to_string(),
                item["value"].as_str().unwrap_or_default().to_string(),
            ))
            .or_default()
            .push(json!({ "file": item["file"], "line": item["line"] }));
    }
    let mut duplicates = Vec::new();
    for ((_repository, identifier), entries) in locations {
        let files: BTreeSet<&str> = entries
            .iter()
            .filter_map(|entry| entry["file"].as_str())
            .collect();
        if files.len() > 1 {
            duplicates.push(json!({ "kind": "duplicate-identifier", "identifier": identifier, "locations": entries }));
        }
    }
    let mut errors = duplicates;
    errors.extend(missing);
    errors.extend(forbidden);
    let referenced: BTreeSet<String> = references
        .iter()
        .filter_map(|item| item["value"].as_str().map(str::to_string))
        .collect();
    let mut unused: Vec<String> = modifiers
        .iter()
        .filter_map(|item| item["value"].as_str())
        .filter(|identifier| {
            identifier.starts_with(&app_prefix) && !referenced.contains(*identifier)
        })
        .map(str::to_string)
        .collect::<BTreeSet<_>>()
        .into_iter()
        .collect();
    unused.sort();
    Ok(json!({
        "schemaVersion": 1,
        "appId": app_id,
        "ok": errors.is_empty(),
        "sourceFiles": scanned_files,
        "specFiles": spec_files.iter().map(|file| file.to_string_lossy().into_owned()).collect::<Vec<_>>(),
        "identifiers": { "defined": defined.len(), "explicit": modifiers.len(), "referenced": referenced.len(), "unused": unused },
        "errors": errors,
    }))
}

pub fn accessibility_command(harness: &Path, app_id: &str) -> Result<bool, Failure> {
    let result = validate_accessibility(harness, app_id)?;
    let ok = result["ok"].as_bool().unwrap_or(false);
    print_json(&result)?;
    Ok(ok)
}

